framework/src/utils/SplineInterpolationBase.C

```diff
@@ -0,0 +1,122 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "SplineInterpolationBase.h"
+#include "MooseError.h"
+
+SplineInterpolationBase::SplineInterpolationBase()
+{}
+
+void
+SplineInterpolationBase::spline(const std::vector<Real> & x, const std::vector<Real> & y, std::vector<Real> & y2, Real yp1/* = 1e30*/, Real ypn/* = 1e30*/)
+{
+  unsigned int n = x.size();
+  if (n == 0)
+    mooseError("Solving for an empty SplineInterpolation.");
+
+  std::vector<Real> u(n, 0.);
+  y2.resize(n, 0.);
+
+  if (yp1 >= 1e30)
+    y2[0] = u[0] = 0.;
+  else
+  {
+    y2[0] = -0.5;
+    u[0] = (3.0 / (x[1] - x[0])) * ((y[1] - y[0]) / (x[1] - x[0]) - yp1);
+  }
+  // decomposition of tri-diagonal algorithm (y2 and u are used for temporary storage)
+  for (unsigned int i = 1; i + 1 < n; i++)
+  {
+    Real sig = (x[i] - x[i - 1]) / (x[i + 1] - x[i - 1]);
+    Real p = sig * y2[i - 1] + 2.0;
+    y2[i] = (sig - 1.0) / p;
+    u[i] = (y[i + 1] - y[i]) / (x[i + 1] - x[i]) - (y[i] - y[i - 1]) / (x[i] - x[i - 1]);
+    u[i] = (6.0 * u[i] / (x[i+1] - x[i - 1]) - sig * u[i - 1]) / p;
+  }
+
+  Real qn, un;
+  if (ypn >= 1e30)
+    qn = un = 0.;
+  else
+  {
+    qn = 0.5;
+    un = (3.0 / (x[n - 1] - x[n - 2])) * (ypn - (y[n - 1] - y[n - 2]) / (x[n - 1] - x[n - 2]));
+  }
+
+  y2[n - 1] = (un - qn * u[n - 2]) / (qn * y2[n - 2] + 1.);
+  // back substitution
+  for (int k = n - 2; k >= 0; k--)
```

Comment:
Implicit sign conversion here. You probably did this so that you wouldn't get a nasty error message about using an unsigned int to compare greater than zero. You can use an unsigned type if you write your for look like this:

```
mooseAssert (n >= 2, "n is too small");
for (unsigned int k = n - 2; k; k--)
...
```

Is *n* always greater than 2? If not this new code probably isn't any better.